src/architecture/serialization.md

```diff
@@ -1,533 +1,193 @@
-used to look up the implementation of [the transaction interface](transactions.md#interface)
-(e.g., `verify` and `execute` methods).
+## Building Exonum with Protobuf Serialization
 
-**Binary presentation:** `u16` (unsigned 2-byte integer).  
-**JSON presentation:** number.
+Exonum includes the
+[`exonum_build`](https://github.com/exonum/exonum/tree/master/exonum_build)
+crate which lets users add the `protobuf_generate` function to their
+`build.rs`. This function automatically generates the `.rs` files for all the
+`.proto` files during the build process. To use `protobuf_generate`, add the
+following code to `build.rs` indicating the folder which contains the `.proto`
+files:
 
-### Message ID
-
-`message_id` defines the type of message within the service.
-
-!!! note "Example"
-    [The sample cryptocurrency service][cryptocurrency] includes 2 main
-    types of transactions: `TxCreateWallet` for creating a wallet with an initial
-    coins balance,
-    and `TxTransfer` for coin transfer.
-
-**Binary presentation:** `u16` (unsigned 2-byte integer).  
-**JSON presentation:** number.
+```rust
+use exonum_build::protobuf_generate;
 
-### Payload length
+protobuf_generate("src/proto", &["src/proto"], "example_mod.rs")
+```
 
-The length of the entire message serialization (including its header and signature).
+To use protobuf-generated Rust structures, users first need to create a module
+which will include the protobuf-generated files:
 
-**Binary presentation:** `u32` (unsigned 4-byte integer).  
-**JSON presentation:** (not serialized).
+```rust
+extern crate exonum;
 
-### Body
+include!(concat!(env!("OUT_DIR"), "/example_mod.rs"));
 
-Serialized [structure](#structures) (including its header and body) described on
-`transactions!` macro call.
+// If you use types from `exonum` .proto files.
+use exonum::proto::schema::*;
+```
 
-### Signature
+Then to access a structure from a protobuf-generated file, you need to indicate
+the name of the module with the generated files and the name of the required
+file and structure in it:
 
-[Ed25519 digital signature](https://ed25519.cr.yp.to/) over the binary
-serialization of the message (excluding the signature bytes,
-i.e., the last 64 bytes of the serialization).
+```rust
+module_name::proto_file_name::StructNameInProtoFile
+```
 
-**Binary presentation:** Ed25519 signature (64 bytes).  
-**JSON presentation:** hex string.
+For example, the generated `Wallet` structure, included in the
+`cryptocurrency.proto` file, which resides in the [`proto`][module] module,
+will be available using `proto::cryptocurrency::Wallet`.
 
-### Example of `transactions!` Usage
+`exonum_build` also includes a function which returns the path to the `.proto`
+files that come with the `exonum` crate - `get_exonum_protobuf_files_path()`:
 
 ```rust
-const MY_SERVICE_ID: u16 = 777;
-
-transactions! {
-    Transactions {
-        const SERVICE_ID = MY_SERVICE_ID;
+let exonum_protos = get_exonum_protobuf_files_path();
```

Comment:
@dvush This is off-topic, but I can't help but notice that this UX seems overly complicated.

- Are there cases in which we *lose* something from including Exonum protobuf files? If no, it could make sense to make their inclusion implicit.
- Duplicating `"src/proto"` dir looks weird.
- It could probably make sense to have a convention for the output module name if we assume `protobuf_generate` will be called once per crate.
- I think it could make sense to organize `proto` files inside the included dir as `exonum/somefile.proto`, analogous to files with well-known Protobuf types (cf. `import "google/protobuf/any.proto";`).

Reply:
4) I agree, and I had to [go back](https://github.com/exonum/exonum/pull/1088/commits/97a9a07be88e33fde820cfd3600e2b216e685edd) to the current version since I could not figure out how to do this on windows, `protoc_generate` uses strings to compare paths in its internals and it breaks on windows because of slashes. I know that it sounds stupid but I could not figure out easy solution to this(and we are bound by time =\)

1,2) I tried to make API similar to `protoc_generate` crate + it would make things more implicit(maybe easier to debug), but now I would agree that it makes more sense to do this implicitly.

3) I prefer current interface since it explicitly ties `include!` and `protoc_generate` (it is greppable and more transparent) and because we have examples where it is called multiple times per build rs(exonum and time-service).

Reply:
Ok, 4) is going to be fixed soon.

Reply:
- Re 1,2: Including exonum files could be an opt-out option if we worry about that.
- Re 3: Wouldn't it be an option to create a macro (say, `include_protobuf!()`) that invokes `include!(concat!(env!("OUT_DIR"), "/mod.rs"))`? It could be used to import exonum-Protobuf glue as well 🙃